src/AccountIdEndpointMiddleware.php

```diff
@@ -0,0 +1,79 @@
+<?php
+namespace Aws;
+
+use Aws\Exception\CredentialsException;
+
+/**
+ * This middleware class resolves the identity from a credentials provider callable function
+ * and determine whether an account should have been resolved. When this middleware resolves
+ * identity then, the identity is included in the $command context bag property "$command['@context]"
+ * as resolved_identity. Example $command['@context]['resolved_identity'] = $resolvedIdentity, and
+ * when this property is set then, the signer middleware gives preference to use that resolved identity
+ * instead of resolving the provided credentials provider by the client. This is done to avoid having to
+ * resolve credentials more than once per request.
+ */
+class AccountIdEndpointMiddleware
+{
+    /**
+     * @var callable $nextHandler
+     */
+    private $nextHandler;
+    /**
+     * @var string $accountIdEndpointMode
+     */
+    private $accountIdEndpointMode;
+    /**
+     * @var callable $credentialsProvider
+     */
+    private $credentialsProvider;
+
+    /**
+     * @param callable $nextHandler
+     * @param string $accountIdEndpointMode
+     * @param callable $credentialsProvider
+     */
+    public function __construct($nextHandler, $accountIdEndpointMode, $credentialsProvider)
+    {
+        $this->nextHandler = $nextHandler;
+        $this->accountIdEndpointMode = $accountIdEndpointMode;
+        $this->credentialsProvider = $credentialsProvider;
+    }
+
+    /**
+     * This method wraps a new instance of the AccountIdEndpointMiddleware.
+     *
+     * @param string $accountIddEndpointMode
+     * @param callable $credentialsProvider
+     * @return callable
+     */
+    public static function wrap($accountIddEndpointMode, $credentialsProvider): callable
+    {
+        return function (callable $handler) use ($accountIddEndpointMode, $credentialsProvider) {
+            return new self($handler, $accountIddEndpointMode, $credentialsProvider);
+        };
+    }
+
+    public function __invoke($command)
+    {
+        $nextHandler = $this->nextHandler;
+        $fnCredentialsProvider = $this->credentialsProvider;
+        $resolvedIdentity = $fnCredentialsProvider()->wait();
+        if (empty($resolvedIdentity->getAccountId())) {
```

Comment:
This should be simplified.  We should use shorter, more concise and separate messages for the exception and warning cases and leave the finer details to doc annotations.  For raising a warning, we use `trigger_error`